lib/uplink/ext/uplink.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package main
+
+// #cgo CFLAGS: -g -Wall
+// #ifndef STORJ_HEADERS
+//   #define STORJ_HEADERS
+//   #include "c/headers/main.h"
+// #endif
+import "C"
+import (
+	"context"
+	"gopkg.in/spacemonkeygo/monkit.v2"
+
+	"storj.io/storj/lib/uplink"
+)
+
+var mon = monkit.Package()
+
+//export NewUplink
+func NewUplink(cErr **C.char) (cUplink C.UplinkRef_t) {
+	goUplink, err := uplink.NewUplink(context.Background(), &uplink.Config{})
+	if err != nil {
+		*cErr = C.CString(err.Error())
+		return cUplink
+	}
+
+	return C.UplinkRef_t(structRefMap.Add(goUplink))
+}
+
+//export NewUplinkInsecure
+func NewUplinkInsecure(cErr **C.char) (cUplink C.UplinkRef_t) {
```

Comment:
why creating new constructors and not just create struct on C side with the same fields? later just reassign them to original struct

Reply:
This is an example where we need to keep a reference to a specific go struct instance. The go struct returned from `NewUplink` has methods on it and no exported fields, so there's no need for a c struct of that type and instead we use the go struct map to store that struct instance where `cUplink` is the key to that value in the map.

Reply:
Sorry, I my comment was unclear. I was talking about passing `uplink.Config` C binding version and have only one `NewUplink`. Something like ```func NewUplink(cUplinkConfig *C.UplinkConfig, cErr **C.char) (cUplink C.UplinkRef_t) {```

Reply:
Oh, yes. We would do that but the entire uplink config is inside of a `Volatile` field which we're not supposed to reference in the bindings. So, for now, the default config is always used because `nil` is being passed for the config. This unfortunately means that in the testing context we can't change the peer ca whitelist option to make things work with test planet with the production `NewUplink` function. That's why this function is here. Blame @jtolds :p.